checkstyle/src/main.rs

```diff
@@ -0,0 +1,212 @@
+use clap::Parser;
+use regex::Regex;
+use std::fs;
+use std::path::{Path, PathBuf};
+use std::process;
+
+#[derive(Parser, Debug)]
+#[command(author, version, about, long_about = None)]
+struct Args {
+    #[arg(short, long)]
+    target_path: String,
+
+    #[arg(short, long)]
+    regex_file_path: String,
+
+    #[arg(short, long)]
+    include: Option<Vec<String>>,
+    #[arg(short, long)]
+    exclude: Option<Vec<String>>,
+}
+
+fn main() -> Result<(), std::io::Error> {
+    let args = Args::parse();
+
+    let regex_file_path = PathBuf::from(&args.regex_file_path);
+    let target_path = PathBuf::from(&args.target_path);
+
+    let include_patterns = args.include;
+    let exclude_patterns = args.exclude;
+
+    let nonmatching_files = separate_regex_matching_files(
+        &regex_file_path,
+        &target_path,
+        include_patterns,
+        exclude_patterns,
+    );
+
+    for file in &nonmatching_files {
+        println!(
+            "File {} does not match regex: {}",
+            file.display(),
+            regex_file_path.display()
+        );
+    }
+
+    if !nonmatching_files.is_empty() {
+        process::exit(1);
+    }
+
+    Ok(())
+}
+
+fn read_file_content(file_path: &Path) -> Option<String> {
+    match fs::read_to_string(file_path) {
+        Ok(content) => Some(content),
+        Err(err) => {
+            if err.kind() == std::io::ErrorKind::InvalidData {
+                eprintln!(
+                    "Warning: File '{}' contains invalid UTF-8 data, skipping...",
+                    file_path.display()
+                );
+                None
+            } else {
+                eprintln!("Failed to read file '{}': {}", file_path.display(), err);
+                None
+            }
+        }
+    }
+}
+
+fn check_matching(file: &str, regex: &str) -> bool {
+    let re = Regex::new(&format!("(?m){}", regex)).unwrap();
+    re.is_match(file)
+}
+
+fn list_files_in_folder(folder_path: &Path) -> Result<Vec<PathBuf>, std::io::Error> {
+    let mut file_paths = Vec::new();
+
+    for entry in fs::read_dir(folder_path)? {
+        let entry = entry?;
+        let path = entry.path();
+
+        if path.is_symlink() {
+            continue;
+        }
+
+        if path.is_file() {
+            file_paths.push(path.clone());
+        } else if path.is_dir() {
+            if let Ok(sub_paths) = list_files_in_folder(&path) {
+                file_paths.extend(sub_paths);
+            }
+        }
+    }
+
+    Ok(file_paths)
+}
+
+fn all_files(folder_path: &Path) -> Vec<PathBuf> {
+    list_files_in_folder(folder_path).unwrap_or_default()
+}
+
+fn separate_regex_matching_files(
```

Comment:
I think this function does a little too much. It would be good for us to separate this out into at least two functions: 
* one that takes the patterns and the target path as arguments and returns a list of paths
* another that takes a list of paths and the regex file path and checks those for validity